Read ClusterAuthToken's hash from  the corresponding Secret if present

andreas-kupries opened the pull request:
ref https://github.com/rancher/rancher/issues/49473
ref https://github.com/rancher/rancher/pull/49493

pmatseykanets reviewed (commented):
Can we make this change backwards compatible? If the secret wasn't yet created for the corresponding ClusterAuthToken use the original logic.

pmatseykanets commented:
We'll need to rebase this on top of #36 once it's merged.

andreas-kupries committed:
beware: go.mod rancher is pointed to rancher PR branch holding upstream and core verification changes
chore: implement new auth logic for cluster auth tokens (cat) and secrets (catas)
note: especially migration of single CATs to CAT+CATAS on auth

andreas-kupries commented:
> We'll need to rebase this on top of #36 once it's merged.

Rebased.

andreas-kupries committed:
fix: linter warnings

pmatseykanets commented on pkg/service/handlers/v1-kube-api-authn.go:
```diff
@@ -118,11 +114,43 @@ func (h *KubeAPIHandlers) v1getAndVerifyUser(accessKey, secretKey string) (*type
+		// Update shadow token to complete the migration
+		if err == nil {
+			clusterAuthToken.SecretKeyHash = "" // nolint:staticcheck
+			_, _ = h.clusterAuthTokens.Update(clusterAuthToken)
```
We should at least log the error. Also do we need to retry on Conflict as in `retry.RetryOnConflict`?

andreas-kupries replied:
Logging :ok: Retry ... Hm ... This in the migration path. Note the comment later, i.e.
>  // Note: migration errors leave a state behind where the
		// migration can be attempted again, either by the next auth
		// request, or the upstream token controller.

andreas-kupries replied:
commit 3002e7c20afcd043e4b5138466a9da554195c6a9

pmatseykanets commented on pkg/service/handlers/v1-kube-api-authn.go:
```diff
@@ -118,11 +114,43 @@ func (h *KubeAPIHandlers) v1getAndVerifyUser(accessKey, secretKey string) (*type
+		clusterAuthTokenSecret, err = h.secrets.Create(clusterAuthTokenSecret)
+		if err != nil && apierrors.IsAlreadyExists(err) {
+			// ... Overwrite an existing secret.
+			_, err = h.secrets.Update(clusterAuthTokenSecret)
```
We should log the error. Do we need to retry on Conflict (though this is less likely to happen than with ClusterAuthTokens)?

andreas-kupries replied:
See above.

andreas-kupries replied:
commit 3002e7c20afcd043e4b5138466a9da554195c6a9

andreas-kupries committed:
address comment: log migration errors.

pmatseykanets commented on pkg/service/handlers/v1-kube-api-authn.go:
```diff
@@ -97,58 +100,58 @@ func v1getBodyAuthnRequest(bytes []byte) (*types.V1AuthnRequest, error) {
-	}
+	var (
+		clusterAuthToken       *clusterv3.ClusterAuthToken
+		clusterAuthTokenSecret *corev1.Secret
```
It seems it's not used outside of the `RetryOnConflict` scope.

andreas-kupries replied:
reworked, commit a0fdbf0f6065ae35ca34961264d9617f35772523

pmatseykanets commented on pkg/service/handlers/v1-kube-api-authn.go:
```diff
@@ -188,11 +191,39 @@ func (h *KubeAPIHandlers) v1getAndVerifyUser(accessKey, secretKey string) (*type
+		return nil, nil, nil, fmt.Errorf("user is not enabled")
+	}
+
+	// An is-not-found error is ok. Likely a not-yet-migrated cluster auth token.
+	// Everything else is reported.
```
IMHO a self-evident comment
```suggestion
```

andreas-kupries replied:
reworked, commit a0fdbf0f6065ae35ca34961264d9617f35772523

pmatseykanets commented on pkg/service/handlers/v1-kube-api-authn.go:
```diff
@@ -97,58 +100,58 @@ func v1getBodyAuthnRequest(bytes []byte) (*types.V1AuthnRequest, error) {
+	)
+	err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		var err error
+		clusterAuthToken, clusterAuthTokenSecret, clusterUserAttribute, err = h.v1GetClusterAuth(accessKey)
+		if err != nil {
+			return err
+		}
```
TBH not a big fan of indirects. I'd rather read top-down what we're doing here instead jumping between methods.

andreas-kupries replied:
reworked, commit a0fdbf0f6065ae35ca34961264d9617f35772523

Me, on the other hand,  not a fan of long functions which do not fit fully on the screen.

andreas-kupries committed:
on conflicts immediately retry migration

andreas-kupries commented:
Tested with a DO cluster. Works. Likely no retry triggered. Testing the retry ... No idea.

pmatseykanets commented on pkg/service/handlers/v1-kube-api-authn.go:
```diff
@@ -37,7 +38,6 @@ func (h *KubeAPIHandlers) v1Authenticate(w http.ResponseWriter, r *http.Request)
 		ReturnHTTPError(w, r, http.StatusBadRequest, fmt.Sprintf("%v", err))
 		return
 	}
```
Probably worth leaving at least a debug level log line
```suggestion
	}
	
	log.Debugf("Processing v1Authenticate request for %s", accessKey)
```

pmatseykanets commented on go.mod:
```diff
@@ -8,68 +8,73 @@ replace (
-	github.com/rancher/rancher/pkg/apis => github.com/rancher/rancher/pkg/apis v0.0.0-20250331174853-4b8ab8c48d68
-	github.com/rancher/rancher/pkg/client => github.com/rancher/rancher/pkg/client v0.0.0-20250331174853-4b8ab8c48d68
+	// reference commit 0dc35390907b on https://github.com/rancher/rancher/pull/49493 - 20250724133739
+	github.com/rancher/rancher => github.com/rancher/rancher v0.0.0-20250724133739-0dc35390907b
+	github.com/rancher/rancher/pkg/apis => github.com/rancher/rancher/pkg/apis v0.0.0-20250724133739-0dc35390907b
+	github.com/rancher/rancher/pkg/client => github.com/rancher/rancher/pkg/client v0.0.0-20250724133739-0dc35390907b
```
These will be replaced with the proper merge commit reference, right?

andreas-kupries replied:
Yes.

andreas-kupries committed:
Update pkg/service/handlers/v1-kube-api-authn.go

Co-authored-by: Peter Matseykanets <pmatseykanets@gmail.com>

andreas-kupries committed:
fix whitespace formatting

andreas-kupries committed:
move r/r references to the #49493 merge commit.
